liquidity-mining/src/lib.rs

```diff
@@ -0,0 +1,1656 @@
+//!     owner. Deleted global farm stay in the storage until all yield farms are removed from
+//!     the storage. Last yield farm removal from storage triggers global farm removal from
+//!     storage.
+//!     Note: deleted global farm CAN'T be resumed.
+//! * Pot - account holding all rewards allocated from all `GlobalFarm`s for all `YieldFarm`s.
+//! `GlobalFarm` transfers allocated rewards to pot account and rewards claimed by `YieldFarm` are
+//! transferred from this account.
+//!
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(test)]
+mod tests;
+mod types;
+
+pub use pallet::*;
+
+pub use crate::types::{
+    Balance, DepositData, DepositId, FarmId, FarmMultiplier, FarmState, GlobalFarmData, GlobalFarmId, LoyaltyCurve,
+    YieldFarmData, YieldFarmEntry, YieldFarmId,
+};
+use codec::{Decode, Encode, FullCodec};
+use frame_support::{
+    pallet_prelude::*,
+    sp_runtime::{
+        traits::{AccountIdConversion, BlockNumberProvider, MaybeSerializeDeserialize, One, Zero},
+        RuntimeDebug,
+    },
+    PalletId,
+};
+
+use frame_system::pallet_prelude::BlockNumberFor;
+use sp_runtime::ArithmeticError;
+
+use hydra_dx_math::liquidity_mining as math;
+use orml_traits::MultiCurrency;
+use scale_info::TypeInfo;
+use sp_arithmetic::{
+    traits::{CheckedDiv, CheckedSub},
+    FixedPointNumber, FixedU128, Perquintill,
+};
+use sp_std::convert::{From, Into, TryInto};
+
+type AssetIdOf<T, I = ()> = <T as pallet::Config<I>>::CurrencyId;
+type PeriodOf<T> = <T as frame_system::Config>::BlockNumber;
+
+#[frame_support::pallet]
+pub mod pallet {
+    use super::*;
+
+    #[pallet::pallet]
+    #[pallet::without_storage_info]
+    pub struct Pallet<T, I = ()>(PhantomData<(T, I)>);
+
+    #[pallet::hooks]
+    impl<T: Config<I>, I: 'static> Hooks<BlockNumberFor<T>> for Pallet<T, I> {}
+
+    #[pallet::config]
+    pub trait Config<I: 'static = ()>: frame_system::Config + TypeInfo {
+        type Event: From<Event<Self, I>> + IsType<<Self as frame_system::Config>::Event>;
+
+        /// Asset type.
+        type CurrencyId: Parameter + Member + Copy + MaybeSerializeDeserialize + Ord + From<u32>;
+
+        /// Currency for transfers.
+        type MultiCurrency: MultiCurrency<Self::AccountId, CurrencyId = Self::CurrencyId, Balance = Balance>;
+
+        /// Pallet id.
+        #[pallet::constant]
+        type PalletId: Get<PalletId>;
+
+        /// Minimum total rewards to distribute from global farm during liquidity mining.
+        #[pallet::constant]
+        type MinTotalFarmRewards: Get<Balance>;
+
+        /// Minimum number of periods to run liquidity mining program.
+        #[pallet::constant]
+        type MinPlannedYieldingPeriods: Get<Self::BlockNumber>;
+
+        /// The block number provider
+        type BlockNumberProvider: BlockNumberProvider<BlockNumber = Self::BlockNumber>;
+
+        /// Id used to identify amm pool in liquidity mining pallet.
+        type AmmPoolId: Parameter + Member + Clone + FullCodec;
```

Comment:
I believe at this point it is better to do the following as there are lot of clone calls for amm_pool_id.

and you can now remove all the "clone()" calls for amm_pool_id in the code.
```suggestion
        type AmmPoolId: Parameter + Member + Copy + FullCodec;
```